File: cpp_engine/live_monitor.cpp
#include <iostream>
#include <iomanip>
#include "json_to_ast.h"
#include "interpreter.h"
#include "data_loader.h"
#include "decay_detector.h"
#include "live_feed_simulator.h"

int main() {
    std::vector<ASTNodePtr> program = load_program("strategy.json");
    std::vector<PriceBar> bars = load_csv("data/RELIANCE_daily.csv");

    Interpreter interp(program);
    DecayDetector detector(10, 3.0);

    bool in_position = false;
    double entry_price = 0.0;
    std::vector<TradeRecord> trades;
    int bars_processed = 0;
    const int max_bars = 300;

    std::cout << "Starting live monitor -- replaying RELIANCE at 0.05s/bar" << std::endl;
    std::cout << "------------------------------------------------------------" << std::endl;

    LiveFeedSimulator feed(bars, 0.05);

    while (feed.advance() && bars_processed < max_bars) {
        const PriceBar& bar_row = feed.current_bar();
        bars_processed++;

        std::map<std::string, double> bar = {
            {"close", bar_row.close}, {"open", bar_row.open},
            {"high", bar_row.high}, {"low", bar_row.low}
        };

        std::vector<std::string> signals = interp.run_bar(bar);
        bool has_buy = std::find(signals.begin(), signals.end(), "BUY") != signals.end();
        bool has_sell = std::find(signals.begin(), signals.end(), "SELL") != signals.end();

        if (has_buy && !in_position) {
            entry_price = bar_row.close;
            in_position = true;
            std::cout << "[" << bar_row.date << "] LIVE SIGNAL: BUY @ " << entry_price << std::endl;
        }

        if (has_sell && in_position) {
            double exit_price = bar_row.close;
            double pnl_pct = (exit_price - entry_price) / entry_price;
            trades.push_back({bar_row.date, pnl_pct * 100.0, pnl_pct > 0 ? 1 : 0});
            std::cout << "[" << bar_row.date << "] LIVE SIGNAL: SELL @ " << exit_price
                      << "  P&L: " << (pnl_pct * 100.0) << "%" << std::endl;
            in_position = false;

            // Re-check decay after every new trade -- continuous monitoring,
            // not a single end-of-run report.
            std::vector<DecayFlag> flags = detector.detect(trades);
            if (!flags.empty() && flags.back().trade_index == static_cast<int>(trades.size()) - 1) {
                std::cout << "    !! DECAY DETECTED at trade #" << trades.size()
                          << " -- win rate has shifted from baseline" << std::endl;
            }
        }
    }

    std::cout << "------------------------------------------------------------" << std::endl;
    std::cout << "Live monitor stopped. " << bars_processed << " bars processed, "
              << trades.size() << " trades executed." << std::endl;

    return 0;
}

File: live_monitor.py
import pandas as pd
from dsl_lexer import tokenize
from dsl_parser import Parser
from dsl_interpreter import Interpreter
from decay_detector import DecayDetector
from live_feed_simulator import LiveFeedSimulator

STRATEGY_SOURCE = """fast_ma = moving_average(close, 5)
slow_ma = moving_average(close, 20)
buy when fast_ma crosses_above slow_ma
sell when fast_ma crosses_below slow_ma"""


def run_live_monitor(csv_path, delay_seconds=0.05, max_bars=200):
    df = pd.read_csv(csv_path)
    program = Parser(tokenize(STRATEGY_SOURCE)).parse_program()
    interp = Interpreter(program)

    feed = LiveFeedSimulator(df, delay_seconds=delay_seconds)
    detector = DecayDetector(baseline_window=10, threshold=3.0)

    position = None
    trades = []
    bars_processed = 0

    print(f"Starting live monitor — replaying {csv_path} at {delay_seconds}s/bar")
    print("-" * 60)

    while feed.advance() and bars_processed < max_bars:
        row = feed.current_bar()
        bars_processed += 1
        bar = {"close": row["Close"], "open": row["Open"], "high": row["High"], "low": row["Low"]}
        signals = interp.run_bar(bar)

        if "BUY" in signals and position is None:
            position = row["Close"]
            print(f"[{row['Date']}] LIVE SIGNAL: BUY @ {position:.2f}")

        elif "SELL" in signals and position is not None:
            pnl_pct = (row["Close"] - position) / position
            trades.append({"date": row["Date"], "pnl_pct": pnl_pct * 100, "win": 1 if pnl_pct > 0 else 0})
            print(f"[{row['Date']}] LIVE SIGNAL: SELL @ {row['Close']:.2f}  P&L: {pnl_pct*100:+.2f}%")
            position = None

            # Re-run decay detection after every new trade — this is the
            # actual point of live monitoring: continuous re-evaluation,
            # not a one-time end-of-backtest check.
            flags = detector.detect(trades)
            if flags and flags[-1][0] == len(trades) - 1:
                print(f"    ⚠ DECAY DETECTED at trade #{len(trades)} — win rate has shifted from baseline")

    print("-" * 60)
    print(f"Live monitor stopped. {bars_processed} bars processed, {len(trades)} trades executed.")


if __name__ == "__main__":
    run_live_monitor("data/RELIANCE_daily.csv", delay_seconds=0.05, max_bars=300)